### ISSUE TITLE
Keras Dropout layer behaviour in test mode seems to be "do nothing" instead of weight scaling inference rule

### ISSUE THREAD

**Issue by durandg12** [tensorflow/tensorflow#29800]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 10.13.6
- TensorFlow installed from (source or binary): from pip install
- TensorFlow version (use command below): v1.12.1-3259-gf59745a381 2.0.0-beta0
- Python version: v3.6.7:6ec5cf24b7, Oct 20 2018, 03:02:14

**Describe the expected behavior**

The weight scaling inference rule is a rule described in [the deep learning book by Goodfellow et al., section 7.12, page 260](https://www.deeplearningbook.org/contents/regularization.html) which consists in, in test or predict mode, multiplying the weights by 1 minus the dropout rate, "to balance for the fact that more units are active than at training time", to quote [the TF 2 tutorial about overfitting, section "Add dropout"](https://www.tensorflow.org/beta/tutorials/keras/overfit_and_underfit), implying that the rule is applied to the Keras `Dropout` layer.

**Describe the current behavior**

I have found no trace of this scaling down of the weights in the code of the `Dropout` Keras layer. In [the actual code](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/layers/core.py), we can see for example, at the end of the `call` method,
```
output = tf_utils.smart_cond(training,
    dropped_inputs,
    lambda: array_ops.identity(inputs))
```
so if `training == False`, the `Dropout` is just the identity.

My first thought was that I was missing something that was hidden elsewhere, maybe in the `fit`, `evaluate` and `predict` method of class `Model`. But then I read [the transformer tutorial](https://www.tensorflow.org/beta/tutorials/text/transformer#evaluate) where custom `evaluate` and `translate` functions are used instead of the methods of class `Model`, along with some `Dropout` layers. And it seems to confirm that calling a `Dropout` layer with `training == False` does not implement the weight scaling inference rule.

**Dropout as a Wrapper for other layers ?**

On the other hand, as a `Dropout` layer is not linked to any set of weights, it seems difficult to implement the weight scaling inference rule as is. Shouldn't `Dropout` be a subclass of `Wrapper` instead of `Layer`, allowing it to be feeded with a layer with some weights, and then applying the dropout to the outputs of this layer in training mode, while scaling down its weights in test and predict mode ?

**Comment by durandg12** [comment]
After reading [the documentation](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/ops/nn_ops.py) of `tf.python.ops.nn_ops.dropout_v2`, it seems that the weight scaling inference is indeed implemented, in the form of a scaling up of the non-zero inputs at training time.

From input x_i, what `dropout_v2` does is either set it to 0, with probability r (r is the dropout rate), either set it to x_i / (1 - r), with probability 1 - r. So the expected value of the output is x_i. So is the expected value of the `Dropout` layer in testing mode (because it is the identity).

As presentend in Goodfellow et al., the rule consists in setting x_i either to 0, either to x_i in training mode, hence an expected value of (1 - r) * x_i. Thus the need to multiply x_i by (1 - r) in testing and predict mode, to keep the same expected value.

So the `Dropout` layer and the `dropout_v2` do indeed implement the weight scaling inference (up to a multiplying factor but that doesn't matter) and I am closing this issue.
